3385 - Regional staff submission history

jtimpe opened the pull request:
## Summary of Changes
Pull request closes #3385 
- Creates a "read only" version of submission history for OFA Regional Staff users
- Regional Staff can only view submission history of locations assocaited with their region
- Reorganizes cypress test data from django management commands to django fixtures. Creates a cypress test for regional staff workflow

## How to Test
```
cd tdrs-backend && docker-compose up --build
cd tdrs-frontend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Assign your user the `OFA Regional Staff` role. Add one or more regions.
3. Navigate to the Data Files page. Use the STT Combobox to select a location. Verify list of available locations is limited to those associated with your user's region.
4. Submit the search form. Verify the upload form cannot be viewed, only submission history can be viewed. Verify that files cannot be downloaded.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Regional user can view submission history for STTs in their region.
- [x] Regional user can't upload files to any section of TDP. 
- [x] Regional user can't download data files from TDP. 
- [x] Clicking 'Search' automatically shows Submission History. 
- [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [x] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

jtimpe committed:
allow regional users to see submission history

jtimpe committed:
filter datafiles stt combobox to regional stts

jtimpe committed:
add backend permission checks

jtimpe committed:
tests

jtimpe committed:
fix tests

jtimpe committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

jtimpe committed:
separate admin api from admin console requests

jtimpe committed:
change feat file discovery

jtimpe committed:
reorg

jtimpe committed:
create user/datafile fixtures

jtimpe committed:
add regional staff submission history test

reitermb commented:
![image](https://github.com/user-attachments/assets/29b5fb64-d072-4863-acee-8f53ae7d4ee6)
Looking good! Clear scan and expected screenreader behavior

elipe17 committed:
- Updated viewset to not allow Regional users to view all users
- Added migration reducing Regional user access
- Removed tests since regional users can no longer manipulate other users

elipe17 committed:
Merge branch '3385-regional-staff-submission-history' into 3476-regional-staff-perms

raftmsohani commented on docs/Technical-Documentation/cypress-integration-tests.md:
````diff
@@ -14,7 +14,7 @@ All tests added into the `tdrs-frontend/cypress/e2e/` folder will be run against
    ```bash
    cd tdrs-backend
-   docker-compose exec web python manage.py generate_cypress_users
+   docker-compose exec web python manage.py loaddata cypress/users cypress/data_files
````
🥇

raftmsohani reviewed (approved):
LGTM!

elipe17 commented on tdrs-backend/tdpservice/data_files/test/test_api.py:
```diff
@@ -427,25 +435,44 @@ def user(self, regional_user):
-        response = self.post_data_file_file(api_client, regional_data_file_data)
+        """Test OFA Regional Staff cannot download datafiles."""
+        post_client = self.login_as(data_analyst)
+        response = self.post_data_file_file(post_client, regional_data_file_data)
```
I know you didn't write this func, but it has a confusing/awkward name. Are you open to renaming it to something like `post_datafile`?

jtimpe replied:
done!

elipe17 commented on tdrs-backend/tdpservice/users/permissions.py:
```diff
@@ -167,6 +167,11 @@ def has_permission(self, request, view):
         # Checks for existence of `data_files.view_datafile` Permission
         has_permission = super().has_permission(request, view)
 
+        # Regional Staff are not allowed to submit or data files
```
```suggestion
        # Regional Staff are not allowed to submit or download data files
```

elipe17 commented on tdrs-frontend/cypress/support/commands.js:
```diff
@@ -63,16 +63,16 @@ Cypress.Commands.add('adminLogin', () => {
 
 Cypress.Commands.add(
-  'adminApiRequest',
+  'adminConsoleFormRequest',
```
Love this name change to help us remember it is admin form fields in the future!

elipe17 commented on tdrs-frontend/src/selectors/stts.js:
```diff
@@ -0,0 +1,8 @@
+  accountIsRegionalStaff(state)
+    ? selectUser(state)
+        .regions?.map((region) => region.stts)
+        .flat()
```
Can we alphabetize this list?

elipe17 reviewed (approved):
Couple of nits. Otherwise this looks great, LGTM!

jtimpe committed:
Update tdrs-backend/tdpservice/users/permissions.py

Co-authored-by: Eric Lipe <125676261+elipe17@users.noreply.github.com>

jtimpe committed:
alphabetize stt list

jtimpe committed:
change test func name

jtimpe committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

jtimpe committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

elipe17 committed:
Merge pull request #3493 from raft-tech/3476-regional-staff-perms

Reduce Regional Staff Group Permissions

jtimpe committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

jtimpe committed:
fix migration conflict

ADPennington committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

ADPennington committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

jtimpe committed:
hide fra upload form for regional staff

jtimpe commented on tdrs-frontend/src/components/Reports/FRAReports.jsx:
```diff
@@ -802,33 +807,39 @@ const FRAReports = () => {
-                <p className="usa-alert__text">{localAlert.message}</p>
-              </div>
-            </div>
+          {!isRegionalStaff && (
```
@ADPennington since FRA is now merged, i added this conditional so that regional staff users are similarly prevented from uploading FRA reports. other than this conditional, i added `OFA Regional Staff` to one of the auth selectors - that was the extent of the change. cc @elipe17

ADPennington committed:
Merge branch 'develop' into 3385-regional-staff-submission-history

ADPennington commented:
@jtimpe this is in pretty good shape! test notes below. let us know if we need a follow-on for the hyperlink. cc: @lhuxraft @reitermb @victoriaatraft @klinkoberstar @ttran-hub 


- regional staff with access to one region can see submission history for one region of STTs :heavy_check_mark:
- regional staff with access to multiple regions can see history for multiple regions of STTs :heavy_check_mark:
- can download error reports ✔️ 
- cannot download TANF/SSP/Tribal/FRA submitted files :warning:
- cannot upload files :heavy_check_mark:


---
:warning: when regional staff have the FRA feature flag, they can view FRA-related submission history. this is great; my only concern here is that some STTs  who  aren't subject to FRA reporting are also in the combobox. I see that the STT selector is mutually exclusive from the fra feature flag. Not a blocking issue, and I'm not seeing a reason to not add the FRA feature flag to regional staff accounts. We may want to consider cleaning this up in the future. for now:
- worth calling out in some meaningful way in the KC
- exploring this as part of initial research with regional staff. 

![stt_no_fra_regional](https://github.com/user-attachments/assets/f13b764f-71e2-4e0e-821f-a26197a2598a)


:warning: noticing some evidence of FRA files that have a hyperlink. I am not able to download these files, so should we remove the hyperlinks? 

![regional_staff_fra_download_link](https://github.com/user-attachments/assets/2b3fb04e-a0c3-4d9b-9447-1bc1c9bb16c0)

jtimpe committed:
fix regional staff file name display, fra submission history

jtimpe commented:
@ADPennington 

> ⚠️ when regional staff have the FRA feature flag, they can view FRA-related submission history. this is great; my only concern here is that some STTs who aren't subject to FRA reporting are also in the combobox. 

We don't currently distinguish which STTs have FRA access vs which ones don't, which would be necessary to achieve what you're suggesting - i believe that is covered by #3484

> ⚠️ noticing some evidence of FRA files that have a hyperlink. I am not able to download these files, so should we remove the hyperlinks?

Good catch, thank you! Fixed

ADPennington commented:
> > ⚠️ noticing some evidence of FRA files that have a hyperlink. I am not able to download these files, so should we remove the hyperlinks?
> 
> Good catch, thank you! Fixed

confirmed fix

![regional_staff_fra_download_linkfixed](https://github.com/user-attachments/assets/704fdb15-6a5c-4715-aebe-42f927c33e40)

ADPennington reviewed (approved):
🚀

lhuxraft commented:
> > ⚠️ when regional staff have the FRA feature flag, they can view FRA-related submission history. this is great; my only concern here is that some STTs who aren't subject to FRA reporting are also in the combobox.
> 
> We don't currently distinguish which STTs have FRA access vs which ones don't, which would be necessary to achieve what you're suggesting - i believe that is covered by #3484

@ADPennington Just chatted with @jtimpe and I actually don't think that will be addressed in #3484 - I will draft up another ticket to cover this use case though.
